metagraph/src/common/vectors/vector_algorithm.hpp

```diff
@@ -109,6 +110,90 @@ inline void unset_bit(uint64_t *v,
     }
 }
 
+inline uint64_t atomic_fetch(const sdsl::int_vector<> &vector, uint64_t i,
+                             std::mutex &backup_mutex) {
+#ifdef MODE_TI
+    size_t width = vector.width();
+    uint64_t bit_pos = i * width;
+    uint8_t shift = bit_pos & 0x7F;
+
+    if (shift + width <= 128) {
+        const __uint128_t *limb = &reinterpret_cast<const __uint128_t*>(vector.data())[bit_pos >> 7];
+        uint64_t mask = (1llu << width) - 1;
+        return (__atomic_load_n(limb, __ATOMIC_SEQ_CST) >> shift) & mask;
+    }
+
+#endif
+
+    std::lock_guard<std::mutex> lock(backup_mutex);
+    std::atomic_thread_fence(std::memory_order_acquire);
+    return vector[i];
+}
+
+inline uint64_t atomic_fetch_and_add(sdsl::int_vector<> &vector, uint64_t i,
+                                     uint64_t count,
+                                     std::mutex &backup_mutex) {
+    // TODO: support adding negatives
+#ifdef MODE_TI
+    size_t width = vector.width();
+    uint64_t bit_pos = i * width;
+    uint8_t shift = bit_pos & 0x7F;
+
+    if (shift + width <= 128) {
+        __uint128_t *limb = &reinterpret_cast<__uint128_t*>(vector.data())[bit_pos >> 7];
+        __uint128_t mask = ((__uint128_t(1) << width) - 1) << shift;
+        __uint128_t inc = __uint128_t(count) << shift;
+        __uint128_t exp_val;
+        __uint128_t new_val;
+        do {
+            exp_val = *limb;
+            new_val = ((exp_val + inc) & mask) | (exp_val & (~mask));
+        } while (!__sync_bool_compare_and_swap(limb, exp_val, new_val));
+
+        return (exp_val & mask) >> shift;
+    }
+
+#endif
+
+    std::lock_guard<std::mutex> lock(backup_mutex);
+    std::atomic_thread_fence(std::memory_order_acquire);
+    uint64_t old_val = vector[i];
+    vector[i] += count;
+    std::atomic_thread_fence(std::memory_order_release);
+    return old_val;
+}
```

Comment:
TODO: This creates a race condition....another thread could have done an atomic CAS operation in the middle of `vector[i] += count`

Reply:
I've disabled it for now so we can look into it later. Maybe we can use transactional memory intrinsics for this at some point.

Reply:
Unaligned atomic operations for <= 64-bit words are allowed, so this was rewritten accordingly.